### PR TITLE
Add `FalconBackbone`

### PR DESCRIPTION
**Pull request by SamanehSaadat**
This is part of addressing #1372 to add the `Falcon` model to KerasNLP. This PR adds the `FalconBackbone`.

This is still a work in progress.

### CODE REVIEW

File: keras_nlp/models/falcon/falcon_transformer_decoder.py [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+        )
+
+        self._post_attention_layernorm = keras.layers.LayerNormalization(
+            epsilon=self.layer_norm_epsilon
+        )
+        self._post_attention_layernorm.build(decoder_sequence_shape)
+
+        # Feedforward layers.
+        self._dense_h_to_4h = keras.layers.Dense(
+            units=4 * self.hidden_dim,
```
Comment (mattdangerw) [inline keras_nlp/models/falcon/falcon_transformer_decoder.py:64]: We generally just allow a second init parameter here for the inner FF dimension. Is there a reason we can't do that here? More flexible and consistent with other keras-nlp models if we can.

File: keras_nlp/models/falcon/falcon_transformer_decoder.py [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+        self.built = True
+
+    def call(
+        self,
+        decoder_sequence,
+        # attention_cache=None,
+        # attention_cache_update_index=None,
+        training=None,
+    ):
+        # TODO: do we need self._compute_self_attention_mask?
```
Comment (mattdangerw) [inline keras_nlp/models/falcon/falcon_transformer_decoder.py:100]: You will definitely want something that adds a causal attention mask here, otherwise we won't match reference output for any sequence with length > 1. `_compute_self_attention_mask` is probably a good place to start.

File: keras_nlp/models/falcon/falcon_transformer_decoder.py [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+
+    def build(self, decoder_sequence_shape):
+        self.hidden_dim = decoder_sequence_shape[-1]
+        self._input_layernorm = keras.layers.LayerNormalization(
+            epsilon=self.layer_norm_epsilon
+        )
+        self._input_layernorm.build(decoder_sequence_shape)
+
+        # Attention layers.
+        # MHA for 1b model.
```
Comment (mattdangerw) [inline keras_nlp/models/falcon/falcon_transformer_decoder.py:42]: What does this comment mean? Do we need a different attention mechanism for other sizes of models?

File: keras_nlp/models/falcon/falcon_backbone.py [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+        hidden_dim,
+        # rope_scaling_factor,
+        layer_norm_epsilon=1e-5,
+        attention_dropout=0,
+        feedforward_dropout=0,
+        **kwargs,
+    ):
+        # Inputs
+        token_ids = keras.Input(shape=(None,), dtype="int32", name="token_ids")
+        # TODO: check falcon's padding.
```
Comment (mattdangerw) [inline keras_nlp/models/falcon/falcon_backbone.py:37]: Probably this can just mimic other models in KerasNLP. I don't see why the padding input would be different (even if attention is custom).

File: keras_nlp/models/falcon/falcon_backbone.py [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+                "token_ids": token_ids,
+            },
+            outputs=sequence_output,
+            **kwargs,
+        )
+        # All references to `self` below this line
+        self.vocabulary_size = vocabulary_size
+        self.num_layers = num_layers
+        self.num_attention_heads = num_attention_heads
+        self.hidden_dim = hidden_dim
```
Comment (mattdangerw) [inline keras_nlp/models/falcon/falcon_backbone.py:79]: Before tests will pass, you will want a `get_config`, as well as `self.token_embedding = token_embedding_layer`. When we add unit tests later on that will catch this stuff.

File: keras_nlp/models/falcon/falcon_transformer_decoder.py [hunk 6]
```diff
@@ -0,0 +1,145 @@
...
+            rate=self.feedforward_dropout,
+            name="feedforward_dropout",
+        )
+
+        self.built = True
+
+    def call(
+        self,
+        decoder_sequence,
+        # attention_cache=None,
```
Comment (mattdangerw) [inline keras_nlp/models/falcon/falcon_transformer_decoder.py:96]: Caching we could definitely do on a follow up if we want. We will need it for generation, but could leave off first PR to keep simple.

**Review by mattdangerw — COMMENTED** [review]
Thanks! Just left some initial comments

File: keras_nlp/models/falcon/falcon_transformer_decoder.py (hunk 3 above)
Comment (SamanehSaadat) [reply keras_nlp/models/falcon/falcon_transformer_decoder.py:42]: Yes, I think different sizes have different attention mechanisms! 1b is MHA, 7b is MQA and 40b is something else!
Different attention mechanisms are handled [here](https://huggingface.co/tiiuae/falcon-rw-1b/blob/e4b9872bb803165eb22f0a867d4e6a64d34fce19/modeling_falcon.py#L211-L216).
Here are the configs for different sizes: [1b](https://huggingface.co/tiiuae/falcon-rw-1b/blob/e4b9872bb803165eb22f0a867d4e6a64d34fce19/config.json#L24-L25), [7b](https://huggingface.co/tiiuae/falcon-7b/blob/898df1396f35e447d5fe44e0a3ccaaaa69f30d36/config.json#L24-L25), [40b](https://huggingface.co/tiiuae/falcon-7b/blob/898df1396f35e447d5fe44e0a3ccaaaa69f30d36/config.json#L24-L25).

File: keras_nlp/models/falcon/falcon_backbone.py (hunk 5 above)
Comment (SamanehSaadat) [reply keras_nlp/models/falcon/falcon_backbone.py:79]: Added `get_config` and `token_embedding`.

File: keras_nlp/models/falcon/falcon_backbone.py (hunk 4 above)
Comment (SamanehSaadat) [reply keras_nlp/models/falcon/falcon_backbone.py:37]: Thanks! Added padding.

File: keras_nlp/models/falcon/falcon_transformer_decoder.py (hunk 1 above)
Comment (SamanehSaadat) [reply keras_nlp/models/falcon/falcon_transformer_decoder.py:64]: I guess I added it to make it clear to myself but I removed it to be consistent with other models.

File: keras_nlp/models/falcon/falcon_transformer_decoder.py (hunk 6 above)
Comment (SamanehSaadat) [reply keras_nlp/models/falcon/falcon_transformer_decoder.py:96]: Sounds good! I'll add caching in a follow up PR.

File: keras_nlp/models/falcon/falcon_transformer_decoder.py (hunk 2 above)
Comment (SamanehSaadat) [reply keras_nlp/models/falcon/falcon_transformer_decoder.py:100]: Thanks for explaining this! I added `_compute_self_attention_mask` similar to the `llama` implementation for now. I'll make adjustment if needed.

File: keras_nlp/models/falcon/falcon_transformer_decoder.py (hunk 3 above)
Comment (SamanehSaadat) [reply keras_nlp/models/falcon/falcon_transformer_decoder.py:49]: As we discussed, to support different attention mechanism with one layer implementation, I will probably use Grouped Query Attention: https://github.com/keras-team/keras/blob/master/keras/layers/attention/grouped_query_attention.py
